# Importing the Libraries

In [1]:
import pandas as pd
import numpy as np

#for data Standardization
from sklearn.preprocessing import StandardScaler

#for train test split
from sklearn.model_selection import train_test_split

#training model
from sklearn import svm
from sklearn.metrics import accuracy_score

# Data Collection and Analysis

### Loading Dataset
We need to load diabetes.csv into dataframes so that we can work with them in Python.

In [2]:
#Mount GoogleColab with Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
#loading dataset to pandas dataframe
df_diabetes = pd.read_csv('drive/My Drive/diabetes.csv')

We can view the first 5 rows of a dataframe using the `head` method. Similarly, if you wanted to see the last 3, you can use `tail(3)`

In [4]:
#Printing first 5 rows of dataset
df_diabetes.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

## Dataset includes the data of all the female patients

**Pregnencies**: the nimber of time a female get pregnant

**Glucose**: Glucose level of female

**Bloood Pressure**: Blood Pressure level

**Skin Thickness**: It is taken from the triceps(It is a large, thick muscle on the dorsal part of the upper arm), basically it tells that fat is present in particular muscle

**Insulin**: Insulin Level of female

**Body Mass Index**: Calculated by deviding the weight in kilograms divided by height in meters squared

**Diabetes Pedigree Function**: Calculates diabetes likelihood depending on the subject's age and her diabetic family history

**Age**: Age of female

**Outcome**: female has diabetes or not. 1 indicates patient has diabetes and 0 indicates no diabetes

In [ ]:
#Total number of rows and column
df_diabetes.shape

(768, 9)

there are 768 people in a rows and 9 columns represents features or parameters. The last column represents the label needed for prediction.

In [ ]:
#Statistical measure of data
df_diabetes.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [6]:
#to check if there is null value in dataset
df_diabetes.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [ ]:
# Total number of values for diabetic and non-diabetic patients
df_diabetes['Outcome'].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

There are 500 people that are non-diabetic (label 0) and 268 people that have diabetes (label 1)

In [ ]:
df_diabetes.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

Get the mean value for all the features on the basis of diabetic and non-diabetic people

In [ ]:
#seperating the data and labels
X = df_diabetes.drop(columns = 'Outcome' , axis=1)
Y = df_diabetes['Outcome']

Drop is used to exclude a specific column or a row in a dataset. Here we drop a column so axis is eaual to one and if we want to drop a row than axis is equal to 0

for X variable, here we getting all the values from the data except Outcome (label) column
and in Y variable storing all the values of Outcome (label) column

In [ ]:
print(X)

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  


In [ ]:
print(Y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


# Data Standardization
standarddize data in a particular range that helps machine learning model for better prediction



In [ ]:
scaler = StandardScaler()

In [ ]:
#Compute the mean and std to be used for scaling.
scaler.fit(X)

StandardScaler()

In [ ]:
#Transform the data
std_data = scaler.transform(X)

here we are fitting all inconsistent data with standard scaler function and based on standardization we are transforming all the data to a common range

In [ ]:
print(std_data)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [ ]:
#for simplification given this standardize data again to variable X
X = std_data

In [ ]:
print(X)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [ ]:
print(Y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


# Train Test Split

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size = 0.2, stratify=Y, random_state=2)

Here we are taking four vairiables X_train, X_test, Y_train and Y_test this will split data into two parts. Test part is used to train machine learning model and once the model is trained we evaluate the model with the test data(unknown data) so that we predict accuracy of the model

train_test_split function is used to seperate test and train data. Where test_size indicates the percentage of data want for test here is 0.2 which means 20% for test data from the  whole data

In [ ]:
print(X.shape, X_train.shape, X_test.shape)

(768, 8) (614, 8) (154, 8)


There are 768 rows of orignal data out of 614 are going to be used for training data  anf 154 for test data

# Training the Model

In [ ]:
#load support vector machine linear model to classifier
classifier = svm .SVC(kernel='linear')

In [ ]:
#fit training data to the classifier
#training the support vector machine classifier
classifier.fit(X_train, Y_train)

SVC(kernel='linear')

# Model Evaluation

## Accuracy Score
### Finding accuracy score on training data

In [ ]:
X_train_pred = classifier.predict(X_train)
train_data_acc = accuracy_score(X_train_pred, Y_train)

In [ ]:
print( 'Accuracy Score of the training data:', train_data_acc)

Accuracy Score of the training data: 0.7866449511400652


Model is predicting 79% correct predictions

### Finding accuracy score on test data

In [ ]:
X_test_pred = classifier.predict(X_test)
test_data_acc = accuracy_score(X_test_pred, Y_test)

In [ ]:
print( 'Accuracy Score of the test data:', test_data_acc)

Accuracy Score of the test data: 0.7727272727272727


The Model is predicting 77% correct prediction on test data so the model is not over trained(over fit)

# Predictive System

In [ ]:
# input the Data
inp_data = (8,120,0,0,0,30,0.183,38)

In [ ]:
# changing input data to numpy array
np_ary = np.asarray(inp_data)

In [ ]:
# reshape the array because we are predicting one instance
reshape_inp_data = np_ary.reshape(1, -1)

In [ ]:
# stansarsize the input data
stand_data = scaler.transform(reshape_inp_data)

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [ ]:
# prediction of input data
Prediction = classifier.predict(stand_data)

if (Prediction[0] == 0):
    print('The person is non_Diabetic')

else:
    print('The person is Diabetic')


The person is Diabetic
